feat: add dynamic expressions

JuroOravec opened the pull request:
Good news, unlike I [said in the discussion](https://github.com/EmilStenstrom/django-components/discussions/552#discussioncomment-10394843) previously, I actually managed to pull off supporting both `{{ }}` and `{% %}` tags (and also `{# #}`)!

Also, after going over how [django-cotton designed it](https://github.com/wrabit/django-cotton#template-expressions-in-attributes), I feel like their approach is more useful than what I proposed initially.

So initially, I was saying that we should allow only a single tag, and no text.

So while this would work:

```django
{% my_component model="{{ model_name }}" %}
```

This would not

```django
{% my_component model="todos.{{ index }}.name" %}
```

But now I think that the latter is more flexible and more can be achieved with it. So the text is practically evaluated as a regular django template.

So @EmilStenstrom this is the same as [you suggested](https://github.com/EmilStenstrom/django-components/discussions/552#discussioncomment-10403531) with:

```django
{% component 'test' title="{{ person.first_name }} {{ person.last_name }}" / %}
```

But to still make it possible to use the outputs of those nested tags as inputs to component, I added following logic:
- If the string consists of a single tag, and no other text, then the output does not get stringified.

	So, below, the result from `random_int` would be passed to `test` component as `int`
	```django
	{% component 'test' page="{% random_int 10 20 %}" / %}
	```
- However, if there IS extra text, or multiple tags, then the result gets stringified.

	Because there are extra spaces around the tag, this would now pass a string.
	```django
	{% component 'test' page=" {% random_int 10 20 %} " / %}
	```

TODO:
- [x] Update README

Closes https://github.com/EmilStenstrom/django-components/discussions/552

EmilStenstrom commented:
@JuroOravec really nice, I like this design!

JuroOravec commented:
@EmilStenstrom Great to hear! I've updated the README, and so this is now ready for review

EmilStenstrom commented on README.md:
````diff
@@ -1927,6 +1931,103 @@ Other than that, you can use spread operators multiple times, and even put keywo
+  value="{% expr 'input_value if is_enabled else None' %}"
+/ %}
+```
+
````
Maybe we should add a disclaimer here? Something like: "Note: Never use this feature to mix business logic and template logic. Business logic should still be in the template!"

(I really like this feature, component is an extremely powerful tag now!)

EmilStenstrom commented on src/django_components/expression.py:
```diff
@@ -109,6 +176,37 @@ def is_aggregate_key(key: str) -> bool:
 
+# A string that must start and end with quotes, and somewhere inside includes
+# at least one tag. Tag may be variable (`{{ }}`), block (`{% %}`), or comment (`{# #}`).
+DYNAMIC_EXPR_RE = re.compile(
+    r"""^{quote}.*?(?:{var_tag}|{block_tag}|{comment_tag}).*?{quote}$""".format(
+        var_tag=r"(?:\{\{.*?\}\})",
+        block_tag=r"(?:\{%.*?%\})",
+        comment_tag=r"(?:\{#.*?#\})",
+        quote=r"['\"]",
+    )
+)
```
This matches strings starting with one quote type, and ending with another: "hallo'

ChatGPT suggests this fix:

```python
DYNAMIC_EXPR_RE = re.compile(
    r"""^(?P<quote>['"]).*?(?:{var_tag}|{block_tag}|{comment_tag}).*?(?P=quote)$""".format(
        var_tag=r"(?:\{\{.*?\}\})",
        block_tag=r"(?:\{%.*?%\})",
        comment_tag=r"(?:\{#.*?#\})",
    )
)

```

JuroOravec replied:
Very nice! I did check for the quotes mismatch, but separately in the `is_dynamic_expression`. I've updated it to check the quotes inside the regex

EmilStenstrom commented on src/django_components/templatetags/component_tags.py:
```diff
@@ -480,13 +481,27 @@ def mark_kwarg_key(key: str, is_agg_key: bool) -> None:
 
         if is_spread_op:
-            expr = parser.compile_filter(val.token)
+            # Allow to use dynamic expressions with spread operator, e.g.
+            # `..."{{ }}"`
```
This is just crazy, I not sure I think we should include this. Will confuse everyone that sees and expression like this!

EmilStenstrom commented on src/django_components/templatetags/component_tags.py:
```diff
@@ -560,6 +575,88 @@ def _parse_tag_body(parser: Parser, end_tag: str, inline: bool) -> NodeList:
 
 
+def _fix_nested_tags(parser: Parser, token: Token) -> None:
+    block_token = token
```
Is the intent here to just rename the variable? Can't this be done in the def right above?

EmilStenstrom commented on src/django_components/templatetags/component_tags.py:
```diff
@@ -560,6 +575,88 @@ def _parse_tag_body(parser: Parser, end_tag: str, inline: bool) -> NodeList:
+    # Above you can see that the token ends at the end of the NESTED tag,
+    # and includes `{%`. So that's what we use to identify if we need to fix
+    # nested tags or not.
+    has_unclosed_tag = block_token.contents.count("{%") > block_token.contents.count("%}")
```
This wouldn't work for this crazy usage I think: `{% component 'test' "{% lorem var_a="{%}" %}`? Do we care? :)

JuroOravec replied:
Of course we do! Fixed it :)

EmilStenstrom commented on tests/test_expression.py:
```diff
@@ -61,6 +78,316 @@ def test_safe_resolve_dict(self):
+        self.assertEqual(captured["bool_var"], "")
+        self.assertEqual(captured["list_var"], "  ")
+
+        self.assertHTMLEqual(
+            rendered,
+            """
+            <div></div>
+            <div> </div>
+            <div></div>
+            <div> </div>
+            """,
```
Maybe not use assertHTMLEqual here since this is testing stringification and handling of spaces?

EmilStenstrom commented on tests/test_expression.py:
```diff
@@ -61,6 +78,316 @@ def test_safe_resolve_dict(self):
+        self.assertHTMLEqual(
+            rendered,
+            """
+            <div>lorem ipsum dolor</div>
+            <div>lorem ipsum dolor [{'a': 1}]</div>
+            <div>True</div>
+            <div>[{'a': 1}, {'a': 2}]</div>
+            <div>{'a': 3}</div>
```
Maybe not use assertHTMLEqual here since this is testing stringification and handling of spaces?

EmilStenstrom commented on tests/test_expression.py:
```diff
@@ -96,7 +423,7 @@ def get_context_data(
                 var_a
                 ...my_dict
-                ...item
+                ..."{{ list|first }}"
```
Are we reaaaaaallly sure we want see this in people's code?! :) 

The steps to understand this code:
1) "Oh, I didn't know the spread operator worked on strings! Oh wait, that's not a string."
2) "It's a list! And first returns the first item. So we're spreading an item in a list, what's that?"
3) "Oh, the list contains little dicts, and this dict doesn't get stringified because there isn't spaces around it."
4) "So this is actually just "a=1", I get it!"

EmilStenstrom reviewed (commented):
This looks very good! I'm a little worried that the parsing here will break when people start doing crazy things like nesting tags several levels deep, quoting django template tags in strings, and expecting them to work.

JuroOravec commented:
@EmilStenstrom Really good feedback, love how thorough it is!

I addressed the feedback [here](https://github.com/EmilStenstrom/django-components/pull/605/commits/0a473a86ec02262594f57bd2e209653dd176c2b3):
- Replaced `assertHtmlEqual` with `assertEqual`
- Covered the `"{%}"` edge case by also checking for unfinished quotes
- Updated the regex to check for quotes in it
- Added the disclaimer for `"Note: Never use this feature to mix business logic and template logic. Business logic should still be in the template!"`
- Added a test to check behavior of [nested tags in nested tags](https://github.com/EmilStenstrom/django-components/pull/605/files#diff-8200e7404420069f578b2bbd98b34348e75285cb3f9a273b85e488ba6eaeadb1R415)

When it comes to points on the complex code like `"'{{ '{{nested}}' }}'"` or `..."{{ list|first }}"`, I don't mind changing the changing the syntax, but I'd really like to have these features.

The reason is, that this comes really handy for the UI component library, where there are cases like this:

```vue
<MyComponent>
  <template #content="contentProps">

    <OtherComp v-bind="{ ...contentProps, ...attrs }" />

  </template>
</MyComponent>
```

In other words, they take data given by a slot (`contentProps`), and combine it with user-provided data (`attrs`), and then spread it onto `OtherComp` using the spread operator (`v-bind`).

So in Django, the equivalent would look like this:

```django
{% component "my_component" %}
	{% fill "content" data="content_props" %}

		{% component "other_comp" ..."{% expr '{ **content_props, **attrs }' %}" %}

	{% endfill %}
{% endcomponent %}
```

Now, the alternative could be again that I'd define the value to spread on a separate line

```django
{% component "my_component" %}
	{% fill "content" data="content_props" %}

		{% expr '{ **content_props, **attrs }'  as merged_props %}
		{% component "other_comp" ...merged_props %}

	{% endfill %}
{% endcomponent %}
```

But I don't see harm in allowing both at the same time. They are complementary concepts. I also didn't document the combined usage (correct me if wrong), so it's more up to the users to discover that it works. And I now also [updated the error message](https://github.com/EmilStenstrom/django-components/pull/605/commits/4beccb63353d68390e067cf814adc8a106f5726d) for when the spread operator receives a non-dict, to make things clear.

So IMO there are guardrails so that this can be used by power users like me, without promoting bad coding habits on the rest.

EmilStenstrom reviewed (approved):
Looks great, and and understand the use-case you are proposing. Lets go with your suggestion!

JuroOravec commented:
Awesome, thanks!
